hexa/metadata/models.py

```diff
@@ -0,0 +1,143 @@
+
+    @staticmethod
+    def encode_base64_id(id: str, model: str) -> str:
+        combined = f"{id}:{model}"
+        encoded = base64.b64encode(combined.encode("utf-8")).decode("utf-8")
+        return encoded
+
+    @staticmethod
+    def decode_base64_id(encoded_id):
+        decoded_bytes = base64.b64decode(encoded_id)
+        decoded_str = decoded_bytes.decode("utf-8")
+        id, model_type = decoded_str.split(":")
+        return id, model_type
+
+    def can_view_metadata(self, user: User):
+        permission = f"{self._meta.app_label}.view_{self.class_name}"
+        if not user.has_perm(permission, self):
+            raise PermissionDenied
+
+    def can_update_metadata(self, user: User):
+        permission = f"{self._meta.app_label}.update_{self.class_name}"
+        if not user.has_perm(permission, self):
+            raise PermissionDenied
+
+    def can_delete_metadata(self, user: User):
+        permission = f"{self._meta.app_label}.delete_{self.class_name}"
+        if not user.has_perm(permission, self):
+            raise PermissionDenied
+
+    def add_attribute(self, key, value, system):
+        content_type = ContentType.objects.get_for_model(self)
+        created = MetadataAttribute.objects.create(
+            content_type=content_type,
+            object_id=self.pk,
+            key=key,
+            value=value,
+            system=system,
+        )
+        return created
+
+    def add_attribute_if_has_permission(self, user, key, value=None, system=False):
+        self.can_update_metadata(user)
+        return self.add_attribute(key, value, system)
+
+    def update_attribute(self, key, system, value):
+        content_type = ContentType.objects.get_for_model(self)
+        metadata_attr, created = MetadataAttribute.objects.update_or_create(
+            content_type=content_type,
+            object_id=self.pk,
+            key=key,
+            defaults={
+                "value": value,
+                "system": system,
+            },
+        )
+        return metadata_attr
+
+    def update_attribute_if_has_permission(self, user, key, value=None, system=False):
+        self.can_update_metadata(user)
+        return self.update_attribute(key, system, value)
+
+    def delete_attribute(self, key):
+        content_type = ContentType.objects.get_for_model(self)
+        return MetadataAttribute.objects.filter(
+            content_type=content_type, object_id=self.pk, key=key
+        ).delete()
+
+    def delete_attribute_if_has_permission(self, user, key):
+        self.can_delete_metadata(user)
+        return self.delete_attribute(key)
+
+    def get_attributes(self, **kwargs):
+        content_type = ContentType.objects.get_for_model(self)
+        metadata_attr = MetadataAttribute.objects.filter(
+            content_type=content_type,
+            object_id=self.pk,
+            **kwargs,
+        )
+        return metadata_attr
+
+    def get_attributes_if_has_permission(
```

Comment:
I would not add a permission system at this level to get attributes. This can be the responsibility of the caller